src/compiler/crystal/syntax/parser.cr

```diff
@@ -2320,7 +2320,16 @@ module Crystal
       open("array literal") do
         next_token_skip_space_or_newline
         while @token.type != :"]"
-          exps << parse_op_assign_no_control
+          exp_location = @token.location
+          has_splat = false
+          if @token.type == :"*"
+            has_splat = true
+            next_token_skip_space_or_newline
+          end
+
+          exp = parse_op_assign_no_control
+          exp = Splat.new(exp).at(exp_location) if has_splat
```

Comment:
Minor structural suggestion, but I think this code is better with two separate branches. The only shared expression is a single call.

```suggestion
          if @token.type == :"*"
            next_token_skip_space_or_newline
            location = @token.location
            exp = Splat.new(parse_op_assign_no_control).at(location)
          else
            exp = parse_op_assign_no_control
          end
```
Ditto below.